modules/platform/nuxeo-drive-server/nuxeo-drive-mongodb/src/main/java/org/nuxeo/drive/mongodb/MongoDBAuditChangeFinder.java

```diff
@@ -0,0 +1,232 @@
+     * {@code DocumentModel} further fetched from the session using the audit entry doc id is fresh.
+     */
+    @Override
+    @SuppressWarnings("unchecked")
+    public long getUpperBound(Set<String> repositoryNames) {
+        long clusteringDelay = getClusteringDelay(repositoryNames);
+        AuditReader auditService = Framework.getService(AuditReader.class);
+        QueryBuilder queryBuilder = new AuditQueryBuilder().predicate(
+                Predicates.in(LOG_REPOSITORY_ID, repositoryNames));
+        if (clusteringDelay > -1) {
+            // Double the delay in case of overlapping, see https://jira.nuxeo.com/browse/NXP-14826
+            long lastClusteringInvalidationDate = System.currentTimeMillis() - 2 * clusteringDelay;
+            queryBuilder.and(Predicates.lt(LOG_LOG_DATE, lastClusteringInvalidationDate));
+        }
+        queryBuilder.order(OrderByExprs.desc(LOG_ID));
+        queryBuilder.limit(1);
+        var entries = (List<LogEntry>) auditService.queryLogs(queryBuilder);
+
+        if (entries.isEmpty()) {
+            if (clusteringDelay > -1) {
+                // Check for existing entries without the clustering invalidation date filter to not return -1 in this
+                // case and make sure the lower bound of the next call to NuxeoDriveManager#getChangeSummary will be >=
+                // 0
+                List<LogEntry> allEntries = (List<LogEntry>) auditService.nativeQuery("", 1, 1);
+                if (!allEntries.isEmpty()) {
+                    log.debug("Found no audit log entries matching the criterias but some exist, returning 0");
+                    return 0;
+                }
+            }
+            log.debug("Found no audit log entries, returning -1");
+            return -1;
+        }
+        return entries.get(0).getId();
+    }
+
+    @Override
+    protected List<LogEntry> queryAuditEntries(CoreSession session, SynchronizationRoots activeRoots,
+            Set<String> collectionSyncRootMemberIds, long lowerBound, long upperBound, int limit) {
+        MongoDBAuditBackend auditService = (MongoDBAuditBackend) Framework.getService(AuditReader.class);
+        MongoCollection<Document> auditCollection = auditService.getAuditCollection();
+        StringBuilder queryBuilder = new StringBuilder();
+        Map<String, Object> params = new HashMap<>();
+        params.put("repositoryId", session.getRepositoryName());
+        params.put("lowerBound", lowerBound);
+        params.put("upperBound", upperBound);
+
+        if(queryStart == null) {
+            try (InputStream is = getClass().getClassLoader().getResourceAsStream("OSGI-INF/queryStart.json")) {
+                queryStart = IOUtils.toString(is, "UTF-8");
+            } catch (IOException e) {
+                log.error("couldn't read query start template", e);
+            }
+        }
+        queryBuilder.append(queryStart);
```

Comment:
I would very much prefer if we had only one file with the whole JSON query and in the middle a special pattern, maybe like `__BODY__` (or something else that makes sense), that gets replaced by what we want. It will make the file much more readable.

But actually after reading more of the code all this string replacement to in the end get a BSON object is really unnatural, it would be better to directly create a BSON object (using `new Document()` etc). This applies also to methods `getCurrentRootFilteringClause` and `getCollectionSyncRootFilteringClause` which should work at the BSON/Document level, not strings.